### ISSUE TITLE
Associations not reduced on update

### ISSUE THREAD

**Issue by mnaughto** [balderdashy/waterline#646]
When I update a record with associations, with no changes to any of the associated records, the associated records are updated anyway. After looking at the source, it seems as though the intention is to skip updating associated records that do not need updated. (This is discussed in /lib/waterline/model/lib/associationMethods/update.js.) However, the reduction, using /lib/waterline/utils/nestedOperations/reduceAssociations.js fails, since the value of attribute.on is the name of the database key column, *not* the model key field, and in my case, the two names are not the same. It seems that attribute.onKey should be used here, instead, but I don't know if that would break things elsewhere.

**Comment by devinivy** [comment]
Mind pointing to some lines of code?

**Comment by mnaughto** [comment]
It's been a while, and I don't remember the specifics of the issue anymore.

I was referring to the comments of line 41 in update.js and lines 46 and 47 in reduceAssociations.js.

**Comment by dmarcelino** [comment]
@mnaughto, take a look at https://github.com/balderdashy/waterline/issues/964#issuecomment-95626676. I think your best option is to not include any associated records in the `.update()` call if you don't intent to change the associations nor the associated records values.

**Comment by dmarcelino** [comment]
I'm convinced that https://github.com/balderdashy/waterline/issues/964#issuecomment-95626676 holds the answer to this issue. There is an ongoing discussion on #975 about how to mitigate/fix this. Please continue the discussion there. If you feel these are 2 different issues, please reopen. Thanks!

**Comment by devinivy** [comment]
@dmarcelino I'm going to reopen this and self-assign, since I think it's a different issue related specifically to the usage of `Model#save` and not `Collection#update` in general.  Also, a fair amount of snooping around code has already been done, which is great!  Keying in on,
>However, the reduction, using /lib/waterline/utils/nestedOperations/reduceAssociations.js fails, since the value of attribute.on is the name of the database key column, not the model key field, and in my case, the two names are not the same. It seems that attribute.onKey should be used here, instead, but I don't know if that would break things elsewhere.

**Comment by dmarcelino** [comment]
That's cool, I'm glad you picked it up.

**Comment by devinivy** [comment]
The test~~s~~ around reduceAssociations ([here](https://github.com/balderdashy/waterline/blob/master/test/unit/utils/reduceAssociations.js)) is _something_, but that utility just needs much more testing.  To get a quick idea, I implemented @mnaughto's suggestion on one of my own branches.  It seems kosher to me and passes all the waterline tests!  Here's the PR: https://github.com/devinivy/waterline/pull/2.  If you all think it's the right idea, I'll make a proper PR on this repo.

**Comment by dmarcelino** [comment]
PR devinivy#2 looks alright.

Regarding the test, if I understand correctly we would want to check if a given record was updated or not, right? A way to check that may be to use some sort of spy in the `beforeValidate` (inspired by issue https://github.com/balderdashy/waterline/issues/751#issue-50808660), that could tell if a given record was updated or not. Alternatively we could also mock an adapter like in [query.update.nested.js](https://github.com/balderdashy/waterline/blob/master/test/unit/query/query.update.nested.js#L190-L197) and check if values updates were the ones we expected to be.

**Comment by devinivy** [comment]
All good thoughts.  As for a unit test of this script, it just has to make sure that nested belongs-to models are reduced to their primary key value when possible.  That's all that reduce associations does.  If the record is not reduced to its primary key, then it is bound to be re-updated and saved during nested update and/or related processes.  It might even try to be recreated or something crazy like that.

One question floating around my head: are associations reduced twice during `Model#save`?  I see it happens [here](https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/associationMethods/update.js#L65), then the values are passed-off to `Collection#update` and are reduced again [here](https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L80).

**Comment by sailsbot** [comment]
Thanks for posting, @mnaughto.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
